File: backend/app/schemas/docker/compose.py
"""Docker Compose 编排相关 Pydantic Schema。"""

from typing import Literal

from pydantic import BaseModel


class ComposeProjectCreate(BaseModel):
    """创建 Compose 项目请求。"""

    project_name: str
    description: str | None = None
    content: str


class ComposeProjectUpdate(BaseModel):
    """更新 Compose 项目元数据请求。"""

    description: str | None = None
    is_active: bool | None = None


class ComposeEditRequest(BaseModel):
    """编辑 Compose 项目并自动部署请求。"""

    content: str
    comment: str | None = None
    description: str | None = None


class ComposeVersionOut(BaseModel):
    """Compose 版本响应。"""

    id: int
    version_number: int
    content: str
    comment: str | None = None
    is_current: bool
    created_by_user_id: int | None = None
    created_at: str

    class Config:
        from_attributes = True


class ComposeVersionCreate(BaseModel):
    """新增 Compose 版本请求。"""

    content: str


class ComposeStackStatusOut(BaseModel):
    """Compose Stack 运行时状态响应。"""

    status: str
    service_count: int
    running_count: int
    ports: list[str]
    last_action: str | None = None
    last_action_at: str | None = None
    updated_at: str


class ComposeProjectOut(BaseModel):
    """Compose 项目响应。"""

    id: int
    project_name: str
    description: str | None = None
    is_active: bool
    current_version: ComposeVersionOut | None = None
    stack: ComposeStackStatusOut | None = None
    config_files: list[str] | None = None
    working_dir: str | None = None
    created_at: str
    updated_at: str

    class Config:
        from_attributes = True


class ComposeActionRequest(BaseModel):
    """Compose 项目操作请求。"""

    action: Literal["up", "down", "restart"]


class ComposeLogQuery(BaseModel):
    """Compose 日志查询参数。"""

    tail: int = 100
    services: list[str] | None = None

File: frontend/src/api/compose.ts
/**
 * Docker Compose 编排相关 API
 */

import api from "@/lib/api";
import type { ContainerInfo } from "@/api/docker";

/** Compose 版本信息 */
export interface ComposeVersion {
  id: number;
  version_number: number;
  content: string;
  comment: string | null;
  is_current: boolean;
  created_by_user_id: number | null;
  created_at: string;
}

/** Compose Stack 状态 */
export interface ComposeStackStatus {
  status: string;
  service_count: number;
  running_count: number;
  ports: string[];
  last_action: string | null;
  last_action_at: string | null;
  updated_at: string;
}

/** Compose 项目 */
export interface ComposeProject {
  id: number;
  project_name: string;
  description: string | null;
  is_active: boolean;
  current_version: ComposeVersion | null;
  stack: ComposeStackStatus | null;
  config_files: string[] | null;
  working_dir: string | null;
  created_at: string;
  updated_at: string;
}

/** 创建项目请求 */
export interface ComposeProjectCreate {
  project_name: string;
  description?: string | null;
  content: string;
}

/** 更新项目请求 */
export interface ComposeProjectUpdate {
  description?: string | null;
  is_active?: boolean | null;
}

/** 编辑项目并自动部署请求 */
export interface ComposeEditRequest {
  content: string;
  comment?: string | null;
  description?: string | null;
}

/** 操作请求 */
export interface ComposeActionRequest {
  action: "up" | "down" | "restart";
}

/** 部署/操作任务响应 */
export interface ComposeDeployTaskResponse {
  task_id: string;
  project_id: number;
  action: string;
  status: string;
}

/** 查询 Compose 项目列表 */
export function listComposeProjects() {
  return api.get<ComposeProject[]>("/docker/compose").then((r) => r.data);
}

/** 查询单个 Compose 项目详情 */
export function getComposeProject(projectId: number) {
  return api.get<ComposeProject>(`/docker/compose/${projectId}`).then((r) => r.data);
}

/** 创建 Compose 项目（异步，返回任务 ID） */
export function createComposeProject(data: ComposeProjectCreate) {
  return api
    .post<ComposeDeployTaskResponse>("/docker/compose", data)
    .then((r) => r.data);
}

/** 更新 Compose 项目元数据 */
export function updateComposeProject(
  projectId: number,
  data: ComposeProjectUpdate
) {
  return api
    .put<ComposeProject>(`/docker/compose/${projectId}`, data)
    .then((r) => r.data);
}

/** 编辑 Compose 项目并自动部署（异步） */
export function editComposeProject(projectId: number, data: ComposeEditRequest) {
  return api
    .post<ComposeDeployTaskResponse>(`/docker/compose/${projectId}/edit`, data)
    .then((r) => r.data);
}

/** 删除 Compose 项目 */
export function deleteComposeProject(projectId: number) {
  return api.delete(`/docker/compose/${projectId}`).then((r) => r.data);
}

/** 查询 Compose 版本列表 */
export function listComposeVersions(projectId: number) {
  return api
    .get<ComposeVersion[]>(`/docker/compose/${projectId}/versions`)
    .then((r) => r.data);
}

/** 切换 Compose 版本 */
export function rollbackComposeVersion(projectId: number, versionId: number) {
  return api
    .post<ComposeVersion>(
      `/docker/compose/${projectId}/versions/${versionId}/rollback`
    )
    .then((r) => r.data);
}

/** 执行 Compose 操作（up/down/restart，异步） */
export function composeAction(
  projectId: number,
  action: ComposeActionRequest["action"]
) {
  return api
    .post<ComposeDeployTaskResponse>(`/docker/compose/${projectId}/action`, {
      action,
    })
    .then((r) => r.data);
}

/** 查询 Compose 日志 */
export function getComposeLogs(
  projectId: number,
  tail: number,
  services?: string[]
) {
  return api
    .get<{ logs: string }>(`/docker/compose/${projectId}/logs`, {
      params: {
        tail,
        services: services?.join(","),
      },
    })
    .then((r) => r.data);
}

/** 查询/同步 Compose Stack 状态 */
export function getComposeStatus(projectId: number) {
  return api
    .get<ComposeStackStatus>(`/docker/compose/${projectId}/status`)
    .then((r) => r.data);
}

/** 查询 Compose 项目维护的容器列表 */
export function getComposeProjectContainers(projectId: number) {
  return api
    .get<ContainerInfo[]>(`/docker/compose/${projectId}/containers`)
    .then((r) => r.data);
}
